src/lib/optimizer/strategy/like_replacement_rule.cpp

```diff
@@ -0,0 +1,87 @@
+#include "operators/operator_scan_predicate.hpp"
+#include "statistics/chunk_statistics/chunk_statistics.hpp"
+#include "storage/storage_manager.hpp"
+#include "storage/table.hpp"
+#include "utils/assert.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+std::string LikeReplacementRule::name() const { return "Like Replacement Rule"; }
+
+void LikeReplacementRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(node);
+  // Continue only on predicate nodes
+  if (predicate_node == nullptr) {
+    _apply_to_inputs(node);
+    return;
+  }
+  const auto expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate_node->predicate());
+  // Continue only if expressions predicate condition is of type PredicateCondition::Like
+  if (expression == nullptr || expression->predicate_condition != PredicateCondition::Like) {
+    _apply_to_inputs(node);
+    return;
+  }
+  const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression->left_operand());
+  const auto value_expression = std::dynamic_pointer_cast<ValueExpression>(expression->right_operand());
+  // Continue only if left expressions is column and right expression is value (column LIKE "asdf%")
+  if (column_expression == nullptr || value_expression == nullptr) {
+    _apply_to_inputs(node);
+    return;
+  }
+  const auto value = boost::get<std::string>(value_expression->value);
+  // Filter strings containing "_", as they cannot be reformulated using this approach
+  if (value.find_first_of("_") != std::string::npos) {
+    _apply_to_inputs(node);
+    return;
+  }
+  const auto offset = value.find_first_of("%");
+
+  // Continue only if the string has a % wildcard and ends with the % wildcard ("asdf%")
+  if (offset == std::string::npos || offset <= 0 || value.length() != offset + 1) {
+    _apply_to_inputs(node);
+    return;
+  }
+  // Calculate lower and upper bound of the search pattern
+  const auto lower_bound = value.substr(0, offset);
+  const auto current_character_value = static_cast<int>(lower_bound.at(lower_bound.length() - 1));
+  // Find next value according to ASCII-table
+  if (current_character_value > 0 && current_character_value < 127) {
+    const auto next_character = static_cast<char>(current_character_value + 1);
+    const auto upper_bound = lower_bound.substr(0, offset - 1) + next_character;
+    const auto lower_bound_node = PredicateNode::make(std::make_shared<BinaryPredicateExpression>(
+        PredicateCondition::GreaterThanEquals, column_expression, std::make_shared<ValueExpression>(lower_bound)));
+    const auto upper_bound_node = PredicateNode::make(std::make_shared<BinaryPredicateExpression>(
+        PredicateCondition::LessThan, column_expression, std::make_shared<ValueExpression>(upper_bound)));
+
+    // Store the input and outputs of the like predicate node
```

Comment:
Adding the new nodes below the old predicate node and then removing the old node would be a bit less code